### PR TITLE
Transaction isolation parameter

### PR DESCRIPTION
**Pull request by Ananya2**
### Overview

Solves issue https://github.com/microsoft/mssql-jdbc/issues/2902
PR: https://github.com/microsoft/mssql-jdbc/pull/2900

This PR introduces support for the `defaultTransactionIsolation` connection property, allowing users to specify the initial transaction isolation level for a connection directly via the JDBC URL or through the `SQLServerDataSource` object. This property defines the isolation level applied immediately after the connection is established.

Using human-readable string values makes the configuration self-explanatory and aligns with modern JDBC driver practices. This change simplifies configuration, especially for environments where the connection string is the primary way to manage driver behavior (e.g., connection pools or third-party tools).

The implementation automatically executes the corresponding `SET TRANSACTION ISOLATION LEVEL` command upon connection establishment if the property is provided.

**Supported Values (Case-Insensitive Strings):**

- `READ_UNCOMMITTED`
- `READ_COMMITTED`
- `REPEATABLE_READ`
- `SERIALIZABLE`
- `SNAPSHOT`

### Example Usage
**Via JDBC URL:**

`jdbc:sqlserver://localhost;databaseName=testdb;defaultTransactionIsolation=READ_UNCOMMITTED;`

**Via SQLServerDataSource:**

```java
SQLServerDataSource ds = new SQLServerDataSource();
ds.setDefaultTransactionIsolation("READ_UNCOMMITTED");
Connection con = ds.getConnection();
```

### Testing
Unit tests have been included in:

`src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java`

to verify:
- Correct parsing of the `defaultTransactionIsolation` property from the JDBC URL.
- Validation of isolation level string values (ensuring only the 5 supported constants are accepted).
- Programmatic configuration via updated `SQLServerDataSource` getters and setters (`setDefaultTransactionIsolation` / `getDefaultTransactionIsolation`).
- Error handling for invalid or unsupported string values.

The changes have been verified to compile and pass tests using the JRE 11 profile (`mvn test -Pjre11 -Dtest=TransactionIsolationPropertyTest`).

### Instrumentation
Meaningful logging has been added to `SQLServerConnection`. When the property is set, the driver logs the isolation level being applied at the `FINE` level, facilitating debugging:

```java
logger.log(Level.FINE, "Setting default transaction isolation level from connection property: {0} ({1})", new Object[]{levelName, levelNum});
```
**Also fixes QUOTED_IDENTIFIER and CONCAT_NULL_YIELDS_NULL choices arrays from {OFF, OFF} to {ON, OFF} — copy-paste bug from PR #2618.**

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 1]
```diff
@@ -5528,6 +5566,33 @@ public String getCatalog() throws SQLServerException {
...
+     *
+     * @param name the isolation level name (e.g. "READ_UNCOMMITTED", "SNAPSHOT")
+     * @return the JDBC transaction isolation level constant, or -1 if invalid
+     */
+    private static int mapTransactionIsolationName(String name) {
+        if (name == null) {
+            return -1;
+        }
+        switch (name.toUpperCase(Locale.US)) {
+            case "READ_UNCOMMITTED":
```
Comment (machavan) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:5581]: check if PG uses same option values like these

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java [hunk 2]
```diff
@@ -1115,10 +1125,13 @@ public final class SQLServerDriver implements java.sql.Driver {
                     Integer.toString(SQLServerDriverIntProperty.CONNECT_RETRY_INTERVAL.getDefaultValue()), false, null),
             new SQLServerDriverPropertyInfo(SQLServerDriverStringProperty.QUOTED_IDENTIFIER.toString(),
                     SQLServerDriverStringProperty.QUOTED_IDENTIFIER.getDefaultValue(), false,
-                    new String[] {OnOffOption.OFF.toString(), OnOffOption.OFF.toString()}),
+                    new String[] {OnOffOption.ON.toString(), OnOffOption.OFF.toString()}),
```
Comment (machavan) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java:1128]: why is this toggled ?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java [hunk 3]
```diff
@@ -1115,10 +1125,13 @@ public final class SQLServerDriver implements java.sql.Driver {
...
             new SQLServerDriverPropertyInfo(SQLServerDriverStringProperty.CONCAT_NULL_YIELDS_NULL.toString(),
                     SQLServerDriverStringProperty.CONCAT_NULL_YIELDS_NULL.getDefaultValue(), false,
-                    new String[] {OnOffOption.OFF.toString(), OnOffOption.OFF.toString()}),};
+                    new String[] {OnOffOption.ON.toString(), OnOffOption.OFF.toString()}),
```
Comment (machavan) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java:1131]: same here

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java [hunk 4]
```diff
@@ -246,6 +246,7 @@ protected Object[][] getContents() {
         {"R_accessTokenCallbackPropertyDescription", "A SQLServerAccessTokenCallback object which is used to call a callback method to return an access token."},
         {"R_quotedIdentifierPropertyDescription", "Indicates whether quotedIdentifier property is set."},
         {"R_concatNullYieldsNullPropertyDescription", "Indicates whether concatNullYieldsNull property is set."},
+        {"R_defaultTransactionIsolationPropertyDescription", "The default transaction isolation level applied when the connection is established."},
```
Comment (machavan) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java:249]: R_transactionIsolationPropertyDescription  - to conform to other prop names ?

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    public static void setupTests() throws Exception {
+        // setConnection() is not called here to allow unit tests to run without a real database connection.
+        // Integration tests will check if connectionString is available before proceeding.
+    }
+
+    @Test
+    public void testPropertyInfo() throws SQLException {
+        SQLServerDriver driver = new SQLServerDriver();
+        Properties info = new Properties();
+        String url = "jdbc:sqlserver://localhost;defaultTransactionIsolation=READ_COMMITTED";
```
Comment (machavan) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: no need for URLs in the tests, tests handle them

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java (hunk 4 above)
Comment (Ananya2) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java:249]: The key R_defaultTransactionIsolationPropertyDescription follows the existing convention where keys align with the exact connection property name (e.g., columnEncryptionSetting, multiSubnetFailover, applicationName). Since the property here is defaultTransactionIsolation, keeping the current key maintains consistency.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java (hunk 1 above)
Comment (Ananya2) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:5581]: @machavan - PG uses space-separated values (e.g., READ COMMITTED) and no dedicated connection property, only the options parameter.
Here, underscore-separated values (READ_COMMITTED, etc.) follow JDBC constants and are easier for connection strings (no quoting). SNAPSHOT is SQL Server-specific and not in PG.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java (hunk 2 above)
Comment (advl03) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java:1128]: This is feedback from Copilot during its review of the first version of the PR:

> In DRIVER_PROPERTIES, the available choices for quotedIdentifier are listed as OFF twice. This prevents tools that rely on DriverPropertyInfo.choices from discovering the valid ON value. Update the choices array to include both ON and OFF (matching OnOffOption).

This is the reason why in the new commit you have the following line:
`new SQLServerDriverPropertyInfo(SQLServerDriverStringProperty.QUOTED_IDENTIFIER.toString(),
                    SQLServerDriverStringProperty.QUOTED_IDENTIFIER.getDefaultValue(), false,
                    new String[] {OnOffOption.ON.toString(), OnOffOption.OFF.toString()}),`

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java (hunk 3 above)
Comment (advl03) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java:1131]: Similar to the previous answer. The feedback from copilot said:

> In DRIVER_PROPERTIES, the available choices for concatNullYieldsNull are listed as OFF twice. This should enumerate both ON and OFF so DriverPropertyInfo.choices is accurate for UIs/introspection.

This is why, in the new commit, you have the following line:
`new SQLServerDriverPropertyInfo(SQLServerDriverStringProperty.CONCAT_NULL_YIELDS_NULL.toString(),
                    SQLServerDriverStringProperty.CONCAT_NULL_YIELDS_NULL.getDefaultValue(), false,
                    new String[] {OnOffOption.ON.toString(), OnOffOption.OFF.toString()}),`

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java (hunk 5 above)
Comment (advl03) [reply src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: OK. Fixed in the new commit for PR #2900 : https://github.com/microsoft/mssql-jdbc/pull/2900/changes/03c22b1f64545b90b3bbedd482fcbf5c4a236b1f

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java (hunk 2 above)
Comment (Ananya2) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java:1128]: Good catch, this was a pre-existing copy-paste bug introduced in https://github.com/microsoft/mssql-jdbc/pull/2618. Both choices were OFF instead of ON and OFF. The contributor spotted it while adding the DEFAULT_TRANSACTION_ISOLATION entry. Fixed it in the latest commit.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 6]
```diff
@@ -1561,4 +1561,19 @@ public interface ISQLServerDataSource extends javax.sql.CommonDataSource {
...
      */
     void setConcatNullYieldsNull(String concatNullYieldsNull);
+
+    /**
+     * Sets the default transaction isolation level for the connection.
+     * 
+     * @param level
+     *        A String that contains the transaction isolation level (e.g. "READ_COMMITTED").
+     */
+    void setDefaultTransactionIsolation(String level);
```
Comment (muskan124947) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java:1571]: Can we rename the methods to setTransactionIsolation() and getTransactionIsolation() to maintain consistency with other properties? Most existing methods do not include “default” in their naming.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 7]
```diff
@@ -3782,6 +3799,27 @@ private void setSessionProperties() throws SQLServerException {
                     break;
             }
         }
+
+        // Apply the defaultTransactionIsolation connection property if set.
+        // This log block helps confirm from the logs what isolation level was requested during the handshake.
+        String defaultTxnIsolationProperty = SQLServerDriverStringProperty.DEFAULT_TRANSACTION_ISOLATION.toString();
+        String defaultTxnIsolationValue = activeConnectionProperties.getProperty(defaultTxnIsolationProperty);
```
Comment (muskan124947) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:3806]: mapTransactionIsolationName is called twice — once during validation and again while applying the property.
Since the value is already validated and mapped earlier, can we avoid re-mapping it again and reuse the computed transactionIsolationLevel instead? This would remove redundant work and ensure consistency.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDataSource.java [hunk 8]
```diff
@@ -1859,6 +1859,16 @@ private Object writeReplace() throws java.io.ObjectStreamException {
         return new SerializationProxy(this);
     }
 
+    @Override
+    public void setDefaultTransactionIsolation(String level) {
+        setStringProperty(connectionProps, SQLServerDriverStringProperty.DEFAULT_TRANSACTION_ISOLATION.toString(), level);
```
Comment (muskan124947) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDataSource.java:1864]: Same :: Can we rename the methods to setTransactionIsolation() and getTransactionIsolation() to maintain consistency with other properties? Most existing methods do not include “default” in their naming.

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    public void testCaseInsensitiveProperty() throws SQLException {
+        SQLServerDriver driver = new SQLServerDriver();
+        Properties info = new Properties();
+        String url = connectionString + ";defaultTransactionIsolation=read_Committed";
+
+        DriverPropertyInfo[] propertyInfo = driver.getPropertyInfo(url, info);
+        assertNotNull(propertyInfo);
+
+        boolean found = false;
+        for (DriverPropertyInfo prop : propertyInfo) {
```
Comment (muskan124947) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: can we - Extract a helper method to reduce duplication and improve readability.
```
private DriverPropertyInfo getProperty(DriverPropertyInfo[] props, String name) {
    for (DriverPropertyInfo prop : props) {
        if (name.equalsIgnoreCase(prop.name)) {
            return prop;
        }
    }
    return null;
}
```

```
DriverPropertyInfo prop = getProperty(propertyInfo, "defaultTransactionIsolation");
assertNotNull(prop);
assertEquals("READ_COMMITTED", prop.value);
```

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        ds.setDefaultTransactionIsolation("READ_UNCOMMITTED");
+        assertEquals("READ_UNCOMMITTED", ds.getDefaultTransactionIsolation());
+    }
+
+    @Test
+    public void testDataSourceDefaultProperty() {
+        com.microsoft.sqlserver.jdbc.SQLServerDataSource ds = new com.microsoft.sqlserver.jdbc.SQLServerDataSource();
+        // By default, the property is null because no default value is defined in the property metadata.
+        // It falls back to driver defaults during connection establishment if not specified.
+        assertTrue(ds.getDefaultTransactionIsolation() == null);
```
Comment (muskan124947) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: assertNull(ds.getDefaultTransactionIsolation());

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+    @Test
+    public void testOutOfRangeProperty() {
+        String url = connectionString + ";defaultTransactionIsolation=NUMERIC_VALUE_UNSUPPORTED";
+
+        SQLException exception = assertThrows(SQLException.class, () -> {
+            PrepUtil.getConnection(url);
+        });
+
+        assertTrue(exception.getMessage().matches(TestUtils.formatErrorMsg("R_InvalidConnectionSetting")),
```
Comment (muskan124947) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: Instead of having 3 asserts we can match with exact error
assertTrue(exception.getMessage().contains("R_InvalidConnectionSetting"));

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+    @Test
+    public void testInvalidProperty() {
+        String url = connectionString + ";defaultTransactionIsolation=invalid";
+
+        SQLException exception = assertThrows(SQLException.class, () -> {
+            PrepUtil.getConnection(url);
+        });
+
+        assertTrue(exception.getMessage().matches(TestUtils.formatErrorMsg("R_InvalidConnectionSetting")),
```
Comment (muskan124947) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: Instead of having 3 asserts we can match with exact error
assertTrue(exception.getMessage().contains("R_InvalidConnectionSetting"));

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+                found = true;
+                break;
+            }
+        }
+        assertTrue(found, "Property 'defaultTransactionIsolation' not found in DriverPropertyInfo");
+    }
+
+    @Test
+    public void testDataSourceProperty() {
+        com.microsoft.sqlserver.jdbc.SQLServerDataSource ds = new com.microsoft.sqlserver.jdbc.SQLServerDataSource();
```
Comment (muskan124947) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: multiple places fully classified class name is used - can we just import and use for readability

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+import com.microsoft.sqlserver.jdbc.ISQLServerConnection;
+import com.microsoft.sqlserver.jdbc.SQLServerDriver;
+import com.microsoft.sqlserver.jdbc.TestUtils;
+import com.microsoft.sqlserver.testframework.AbstractTest;
+import com.microsoft.sqlserver.testframework.Constants;
+import com.microsoft.sqlserver.testframework.PrepUtil;
+
+@RunWith(JUnitPlatform.class)
+public class TransactionIsolationPropertyTest extends AbstractTest {
```
Comment (muskan124947) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: Adding edge cases like empty values or whitespace would further strengthen coverage.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java (hunk 6 above)
Comment (Ananya2) [reply src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java:1571]: We can’t rename these: setTransactionIsolation() / getTransactionIsolation() are already defined in java.sql.Connection for live connection state (int-based). Reusing them here would be misleading.

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java (hunk 11 above)
Comment (Ananya2) [reply src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: R_InvalidConnectionSetting is a resource key, not the actual message, so contains() won’t work.
TestUtils.formatErrorMsg() correctly matches the resolved message, and the extra contains() checks validate parameters.

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+            try (Connection con = PrepUtil.getConnection(url)) {
+                assertEquals(expectedConstants[i], con.getTransactionIsolation(),
+                        "Isolation level " + levels[i] + " was not applied correctly.");
+            }
+        }
+    }
+
+    /**
+     * Integration test to verify case-insensitivity when applying to a real connection.
+     */
```
Comment (machavan) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: test that sets defaultTransactionIsolation to say SERIALIZABLE on connection string ... but explicitly overrides it to say READ_UNCOMMITTED using a statement   and then the test verifies this isolation level semantically.

A vice versa of this or some other combination with the test actual validating the effect of the overriden isolation level

Comment (Ananya2) [reply src/test/java/com/microsoft/sqlserver/jdbc/connection/TransactionIsolationPropertyTest.java]: done
